app/models/miq_alert.rb

```diff
@@ -325,16 +345,16 @@ def build_actions
   end
 
   def eval_expression(target, inputs = {})
-    return Condition.evaluate(self, target, inputs) if expression.kind_of?(MiqExpression)
-    return true if expression.kind_of?(Hash) && expression[:eval_method] == "nothing"
+    return Condition.evaluate(self, target, inputs) if miq_expression
+    return true if hash_expression && hash_expression[:eval_method] == "nothing"
 
-    raise "unable to evaluate expression: [#{expression.inspect}], unknown format" unless expression.kind_of?(Hash)
+    raise "unable to evaluate expression: [#{miq_expression.inspect}], unknown format" unless hash_expression
 
-    case expression[:mode]
+    case hash_expression[:mode]
     when "internal" then return evaluate_internal(target, inputs)
     when "automate" then return evaluate_in_automate(target, inputs)
     when "script"   then return evaluate_script
-    else                 raise "unable to evaluate expression: [#{expression.inspect}], unknown mode"
+    else                 raise "unable to evaluate expression: [#{hash_expression.inspect}], unknown mode"
```

Comment:
Maybe not for this PR, but we are mixing evaluation of 2 different expression types into a single method, and that is causing confusion.  I think this would be a *LOT* clearer if the two evaluation strategies were split into methods, with eval_expression branching between those

```ruby
def eval_expression(target, inputs = {})
  case expression
  when MiqExpression then eval_miq_expression(target, inputs = {})
  when Hash          then eval_hash_expression(target, inputs = {})
  else raise "unable to evaluate expression: [#{expression.inspect}], unknown format"
  end
end
```

I think I'm also leaning towards using expression where possible in the places where it has been changed to individually check miq_expression, then hash_expression, then nil.  Checking for each one in turn is sort of the purpose of the expression method itself.

However, in a method like my suggested `eval_hash_expression`, where you *know* you are working with hash_expression, then yes, you would access it directly.

Reply:
@Fryguy This is defiantly the direction we are aiming to. As @imtayadeway pointed in an earlier discussion, having both expression types in the same column was the root problem here, which we hope to handle in this PR (and its related schema change). The next step will be trying to better encapsulate these types and their evaluation, maybe with two new tables each with one column (it will probably require more discussion). However, since this PR blocks some other items we are working on we would like to move forward with the current implementation as an initial solution if you are OK with it.

Reply:
Sounds great @zakiva